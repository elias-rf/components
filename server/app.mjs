import compression from 'compression'
import history from 'connect-history-api-fallback'
import cookieParser from 'cookie-parser'
import cors from 'cors'
import express from 'express'
import * as path from 'path'
import * as url from 'url'
import { config } from './config/index.mjs'
import { authMiddle } from './middles/auth-middle.mjs'
import { loggerMiddle } from './middles/logger-middle.mjs'
import { routes } from './routes/index.mjs'

const __dirname = url.fileURLToPath(new URL('.', import.meta.url))
// const __filename = url.fileURLToPath(import.meta.url)

export const app = express()

app.use(cookieParser())
app.map = (mapRoute, route) => {
  route = route || ''

  function getRoute(path, route) {
    if (path === '/') {
      return '/' + route
    }
    if (route.startsWith('/')) {
      return path + route
    }
    return path + '/' + route
  }

  for (const itemRoute in mapRoute) {
    switch (typeof mapRoute[itemRoute]) {
      // { '/path': { ... }}
      case 'object':
        app.map(mapRoute[itemRoute], getRoute(route, itemRoute))
        break
      // get: function(){ ... }
      case 'function':
        console.log('%s %s', itemRoute, route)
        app[itemRoute](route, mapRoute[itemRoute])
        break
      default:
        console.log('%s %s não reconhecido!', itemRoute, route)
        break
    }
  }
}

app.set('config', config)
app.use(cors({ origin: '*', credentials: true })) // cliente não precisa de estar no mesmo endereço

app.use(
  history({
    verbose: false,
    index: '/index.html',
  })
)
app.use(compression()) // compressão de dados ao enviar para o cliente

app.use(express.static(path.join(__dirname, '../public'))) // pasta pública
app.use(express.json()) // decodifica body.json
app.use(express.urlencoded({ extended: false })) // decodifica query
app.use(loggerMiddle)
app.use(authMiddle)
app.map(routes)
