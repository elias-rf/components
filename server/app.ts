import { config } from '@/config/index.js'
import { authMiddle } from '@/server/middles/auth-middle.js'
import { loggerMiddle } from '@/server/middles/logger-middle.js'
import { routes } from '@/server/routes/index.js'
import compression from 'compression'
import history from 'connect-history-api-fallback'
import cors from 'cors'
import express from 'express'
import * as path from 'path'
import * as url from 'url'

const __dirname = url.fileURLToPath(new URL('.', import.meta.url))
// const __filename = url.fileURLToPath(import.meta.url)

export const app = express()

app.map = (mapRoute: { [key: string]: any }, route: string) => {
  route = route || ''

  function getRoute(path: string, route: string) {
    if (path === '/') {
      return '/' + route
    }
    if (route.startsWith('/')) {
      return path + route
    }
    return path + '/' + route
  }

  for (const itemRoute in mapRoute) {
    switch (typeof mapRoute[itemRoute]) {
      // { '/path': { ... }}
      case 'object':
        app.map(mapRoute[itemRoute], getRoute(route, itemRoute))
        break
      // get: function(){ ... }
      case 'function':
        console.log('%s %s', itemRoute, route)
        app[itemRoute as 'get' | 'put' | 'post' | 'delete'](
          route,
          mapRoute[itemRoute]
        )
        break
      default:
        console.log('%s %s não reconhecido!', itemRoute, route)
        break
    }
  }
}

app.set('config', config)
app.use(cors()) // cliente não precisa de estar no mesmo endereço

app.use(
  history({
    verbose: false,
    index: '/index.html',
  })
)
app.use(compression()) // compressão de dados ao enviar para o cliente

app.use(express.static(path.join(__dirname, '../public'))) // pasta pública
app.use(express.json()) // decodifica body.json
app.use(express.urlencoded({ extended: false })) // decodifica query
app.use(loggerMiddle)
app.use(authMiddle)
app.map(routes)
