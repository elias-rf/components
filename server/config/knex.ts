import { config } from "./index";

import type { Knex } from "knex";

export const optionsPlano: Knex.Config = config.db.plano;

export const optionsFullvision: Knex.Config = config.db.fullvision;

export const optionsOftalmo: Knex.Config = config.db.oftalmo;
