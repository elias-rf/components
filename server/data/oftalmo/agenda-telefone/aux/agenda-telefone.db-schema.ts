export const phonebook = {
  database: 'oftalmo',
  table: 'phonebook',
  primary: ['id'],
  fields: {
    id: {
      table: 'phonebook',
      name: 'id',
      data_type: 'int',
      max_length: 4,
      numeric_precision: 10,
      numeric_scale: null,
      is_nullable: false,
      default_value: null,
      is_primary_key: true,
      is_unique: true,
      has_auto_increment: false,
      foreign_key_table: null,
      foreign_key_column: null,
      is_generated: false,
      generation_expression: null,
    },
    name: {
      table: 'phonebook',
      name: 'name',
      data_type: 'varchar',
      max_length: 200,
      numeric_precision: null,
      numeric_scale: null,
      is_nullable: false,
      default_value: null,
      is_primary_key: false,
      is_unique: false,
      has_auto_increment: false,
      foreign_key_table: null,
      foreign_key_column: null,
      is_generated: false,
      generation_expression: null,
    },
    department: {
      table: 'phonebook',
      name: 'department',
      data_type: 'varchar',
      max_length: 100,
      numeric_precision: null,
      numeric_scale: null,
      is_nullable: true,
      default_value: null,
      is_primary_key: false,
      is_unique: false,
      has_auto_increment: false,
      foreign_key_table: null,
      foreign_key_column: null,
      is_generated: false,
      generation_expression: null,
    },
    email: {
      table: 'phonebook',
      name: 'email',
      data_type: 'varchar',
      max_length: 100,
      numeric_precision: null,
      numeric_scale: null,
      is_nullable: true,
      default_value: null,
      is_primary_key: false,
      is_unique: false,
      has_auto_increment: false,
      foreign_key_table: null,
      foreign_key_column: null,
      is_generated: false,
      generation_expression: null,
    },
  },
}
