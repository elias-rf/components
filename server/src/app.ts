import { join } from "path";
import Koa from "koa";
import bodyParser from "koa-bodyparser";
import serve from "koa-static";
import etag from "koa-etag";
import historify from "koa-historify";
import logger from "koa-pino-logger";

import knexDbs from "./dal/connections";
import router from "./routes";
import authMiddle from "./middleware/auth-middle";

const app = new Koa();

app.use(authMiddle);

app.context.knexDbs = knexDbs;
app.use(logger());
app.use(etag());
app.use(bodyParser());
app.use(serve(join(__dirname, "../../public")));
app.use(router.routes());
app.use(router.allowedMethods());
app.use(
  historify(join(__dirname, "../../public/index.html"), {
    logger: console.log.bind(console),
  })
);

export default app;
