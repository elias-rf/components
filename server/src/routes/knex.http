POST http://localhost:3000/api/knex/oftalmo
Content-Type: application/json

{
  "from":"phonebook",
  "where":[["id","<","105"]],
  "orderBy":[["id","desc"]]
}

###### Login
POST http://localhost:3000/api/login
Content-Type: application/json

{"user":"elias","password":"jkl"}

###### Logout
POST http://localhost:3000/api/logout


### Me
GET http://localhost:3000/api/me

### History HTML5
GET http://localhost:3000/
accept: text/html