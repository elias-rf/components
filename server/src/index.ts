import http from "http";
import { app } from "./app";
import { logger } from "./lib/logger";

async function start() {
  const port = normalizePort(process.env.PORT || "3000");
  const httpServer = http.createServer();

  httpServer.on("error", onError);
  httpServer.on("listening", onListening);
  httpServer.on("request", app.callback());
  await httpServer.listen({ port });

  /**
   * Normalize a port into a number, string, or false.
   */

  function normalizePort(val: string) {
    var port = parseInt(val, 10);

    if (isNaN(port)) {
      // named pipe
      return val;
    }

    if (port >= 0) {
      // port number
      return port;
    }

    return false;
  }

  /**
   * Event listener for HTTP server "error" event.
   */

  function onError(error: any) {
    if (error.syscall !== "listen") {
      throw error;
    }

    var bind = typeof port === "string" ? "Pipe " + port : "Port " + port;

    // handle specific listen errors with friendly messages
    switch (error.code) {
      case "EACCES":
        console.error(bind + " requires elevated privileges");
        process.exit(1);
        break;
      case "EADDRINUSE":
        console.error(bind + " is already in use");
        process.exit(1);
        break;
      default:
        throw error;
    }
  }

  /**
   * Event listener for HTTP server "listening" event.
   */

  function onListening() {
    logger.info(`ðŸš€ Server ready at http://localhost:${port}`);
  }
}
start();
