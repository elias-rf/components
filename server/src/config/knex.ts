import config from "../config/index";

import type { Knex } from "knex";

export const optionsPlano: Knex.Config = {
  client: "mssql",
  connection: {
    host: config.db.plano.connection?.host,
    user: config.db.plano.connection?.user,
    password: config.db.plano.connection?.password,
    database: config.db.plano.connection?.database,
    options: {
      enableArithAbort: true,
      trustServerCertificate: true,
    },
  },
  debug: false,
  useNullAsDefault: true,
};

export const optionsFullvision: Knex.Config = {
  client: "mssql",
  connection: {
    host: config.db.fullvision.connection?.host,
    user: config.db.fullvision.connection?.user,
    password: config.db.fullvision.connection?.password,
    database: config.db.fullvision.connection?.database,
    options: {
      enableArithAbort: true,
      trustServerCertificate: true,
    },
  },
  debug: false,
  useNullAsDefault: true,
};

export const optionsOftalmo: Knex.Config = {
  client: "mssql",
  connection: {
    host: config.db.oftalmo.connection?.host,
    user: config.db.oftalmo.connection?.user,
    password: config.db.oftalmo.connection?.password,
    database: config.db.oftalmo.connection?.database,
    options: {
      trustServerCertificate: true,
      enableArithAbort: true,
      tdsVersion: "7_3_B",
    },
  },
  debug: false,
  useNullAsDefault: true,
};

if (config.app.mock === true) {
  optionsOftalmo.client = "mysql";
  optionsOftalmo.connection = {
    host: "localhost",
    port: 3306,
    user: "sa",
    password: "123",
    database: "OftalmoXPProducao",
  };
  optionsPlano.client = "mysql";
  optionsPlano.connection = {
    host: "localhost",
    port: 3306,
    user: "sa",
    password: "123",
    database: "Visiontech",
  };
  optionsFullvision.client = "mysql";
  optionsFullvision.connection = {
    host: "localhost",
    port: 3306,
    user: "sa",
    password: "123",
    database: "Visiontech",
  };
}
