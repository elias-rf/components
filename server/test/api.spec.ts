import knex from "knex";
import { getTracker, MockClient, Tracker } from "knex-mock-client";
import { afterEach, beforeAll, describe, expect, it, vi } from "vitest";
import { app } from "../src/app";
import { apiRequest, rpcResponseError } from "./aux";

describe("Api", () => {
  let tracker: Tracker;
  vi.mock("../src/dal/connections", () => ({
    connections: {
      plano: knex({ client: MockClient }),
      oftalmo: knex({ client: MockClient }),
      fullvision: knex({ client: MockClient }),
    },
  }));
  const recordNames = {
    agenda_telefone_id: 171,
    nome: "vago78",
    setor: "qualidade",
    email: "vago",
  };
  const recordFields = {
    id: 171,
    name: "vago78",
    department: "qualidade",
    email: "vago",
  };

  beforeAll(() => {
    tracker = getTracker();
  });

  afterEach(() => {
    tracker.reset();
  });

  it("method desconhecido", async () => {
    const rsp = await apiRequest(app, "agendaTelefoneSchemaDesc", {});
    expect(rsp.status).toEqual(200);
    expect(rsp.body).toEqual(rpcResponseError(-32601, "Method not found"));
  });
});
