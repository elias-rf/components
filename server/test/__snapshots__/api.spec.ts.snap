// Vitest Snapshot v1

exports[`Api > indice 1`] = `
{
  "id": 1,
  "jsonrpc": "2.0",
  "result": {
    "mutation": [],
    "query": [],
  },
}
`;
